otherlibs/systhreads/Makefile

```diff
@@ -14,95 +14,129 @@
 #**************************************************************************
 
 include ../../config/Makefile
+
+ROOTDIR=../..
 CAMLRUN ?= ../../boot/ocamlrun
 CAMLYACC ?= ../../boot/ocamlyacc
 
-ROOTDIR=../..
-CAMLC=$(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib \
-      -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix
-CAMLOPT=$(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib \
-        -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix
-MKLIB=$(CAMLRUN) ../../tools/ocamlmklib
+LIBS = -nostdlib -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/$(UNIXLIB)
+
+CAMLC=$(CAMLRUN) $(ROOTDIR)/ocamlc $(LIBS)
+CAMLOPT=$(CAMLRUN) $(ROOTDIR)/ocamlopt $(LIBS)
+MKLIB=$(CAMLRUN) $(ROOTDIR)/tools/ocamlmklib
 COMPFLAGS=-w +33..39 -warn-error A -g -bin-annot -safe-string
 ifeq "$(FLAMBDA)" "true"
 OPTCOMPFLAGS=-O3
 else
 OPTCOMPFLAGS=
 endif
 
-BYTECODE_C_OBJS=st_stubs_b.o
-NATIVECODE_C_OBJS=st_stubs_n.o
+LIBNAME=threads
+
+ifeq "$(UNIX_OR_WIN32)" "unix"
+HEADER = st_posix.h
+else # Windows
+HEADER = st_win32.h
+endif
 
-THREAD_OBJS= thread.cmo mutex.cmo condition.cmo event.cmo threadUnix.cmo
+BYTECODE_C_OBJS=st_stubs_b.$(O)
+NATIVECODE_C_OBJS=st_stubs_n.$(O)
 
-all: libthreads.a threads.cma
+THREADS_SOURCES := thread.ml mutex.ml condition.ml event.ml threadUnix.ml
 
-allopt: libthreadsnat.a threads.cmxa
+THREADS_BCOBJS := $(THREADS_SOURCES:.ml=.cmo)
+THREADS_NCOBJS := $(THREADS_SOURCES:.ml=.cmx)
 
-libthreads.a: $(BYTECODE_C_OBJS)
-	$(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+MLIFILES=thread.mli mutex.mli condition.mli event.mli threadUnix.mli
+CMIFILES=$(MLIFILES:.mli=.cmi)
 
-st_stubs_b.o: st_stubs.c st_posix.h
-	$(BYTECC) -I../../byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) \
-	   -c st_stubs.c
-	mv st_stubs.o st_stubs_b.o
+all: lib$(LIBNAME).$(A) $(LIBNAME).cma $(CMIFILES)
 
+allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
+
+lib$(LIBNAME).$(A): $(BYTECODE_C_OBJS)
+	$(MKLIB) -o $(LIBNAME) $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+
+lib$(LIBNAME)nat.$(A): $(NATIVECODE_C_OBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
 # Dynamic linking with -lpthread is risky on many platforms, so
 # do not create a shared object for libthreadsnat.
-libthreadsnat.a: $(NATIVECODE_C_OBJS)
-	$(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
-
-st_stubs_n.o: st_stubs.c st_posix.h
-	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
-	            $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
-	            -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
-	mv st_stubs.o st_stubs_n.o
+	$(AR) rc $@ $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
+endif
 
-threads.cma: $(THREAD_OBJS)
-	$(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
-	  -cclib -lunix $(PTHREAD_CAML_LINK)
+$(LIBNAME).cma: $(THREADS_BCOBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
```

Comment:
Alain Frisch (2016/08/30 01:08 -0700):
> >  
> > -threads.cma: $(THREAD_OBJS)
> > -	$(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
> > -	  -cclib -lunix $(PTHREAD_CAML_LINK)
> > +$(LIBNAME).cma: $(THREADS_BCOBJS)
> > +ifeq "$(UNIX_OR_WIN32)" "unix"
> > +	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
> > +else # Windows
> > +	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
> 
> Merging makefiles has the nice benefit of making such differences
> apparent, and one should then really either remove or document them.
> But indeed, this can come later.

I agree. For this specific example, it's just that it' was not obvious
to me what to do.

There are other places where I got rid of the differences, e.g. IIRC not
all the files were installed properly on Windows, I think it was the
.mli files which were not installed.

In the same vein, the warning-related flags were also different and I
took the most strict ones.